### DISCUSSION TITLE
PorousFlow: Non-Newtonian modification

### DISCUSSION THREAD

**Discussion by sinaomrani96** [idaholab/moose#25195] [Q&A Modules: General]
Hi,
Hope you are doing well.

Dear @cpgr and @WilkAndy, I am trying to implement non-Newtonian fluid flow in PorousFlow module for my application. The way I could think of is to write my own fluid property file based on "PorousFlowSingleComponentFluid" and then make viscosity as a function of velocity. However, I am facing segmentation fault when I am running it. I was wondering if there is a better way to implement it. Thanks.

**Comment by Joseph-0123** [comment]
Hello,

the segment fault always happens. To check your modification first, you can use the [Debugging](https://mooseframework.inl.gov/application_development/debugging.html).

if you cannot identify the debugging feedback, please paste them here.

Good luck.

**Reply by sinaomrani96** [reply]
Hi,

I have tried the debugging, and change the code. Now I am facing another error:
Cyclic dependency detected in object ordering:
water <- PorousFlow_density_qp_all <- velocity1 <- water

Since the nature of non-Newtonian flow is that the viscosity depending on velocity and velocity has viscosity in itself. So, I guess I need to break this cyclic dependency? I have seen that they did a similar work with porosity in PorousFlow module but I am not sure how I can implement it for viscosity.

**Reply by Joseph-0123** [reply]
Hello, I think I find what you mention about the dynamic [porosity](https://github.com/idaholab/moose/blob/next/modules/porous_flow/src/materials/PorousFlowPorosity.C).

![image](https://github.com/idaholab/moose/assets/56687392/fb6e1f81-cd27-4f50-82bc-d3be831dab7b)

As shown in the above pic, you can use/add _velocity_old or _viscosity_old to break the cyclic dependency between viscosity and velocity.

**Reply by sinaomrani96** [reply]
Hi Joseph,

Thanks it worked by using old velocity. Sorry if my next question is naive, but now I have compatibly issues as follow:

error: cannot convert 'const libMesh::VectorValue<double>' to 'double' in assignment

How I can solve this? Thanks.

**Reply by Joseph-0123** [reply]
I guess you have some coding errors. Please pay attention to the 'vector' you define and refer to the following discussion. 
https://github.com/idaholab/moose/discussions/16139

**Reply by sinaomrani96** [reply]
Thanks for your reply. I solved the issue for the vector definition and now I am back at Segmentation fault error. This is the output for debugging:

Thread 1 "porous_flow-dbg" received signal SIGSEGV, Segmentation fault.
0x00007ffff789c010 in libMesh::TypeVector<double>::operator*<double> (this=0x0, p=...) at /home/sinaomrani/mambaforge3/envs/moose/libmesh/include/libmesh/type_vector.h:886
886	  return (_coords[0]*p(0) +
(gdb) bt
#0  0x00007ffff789c010 in libMesh::TypeVector<double>::operator*<double> (
    this=0x0, p=...)
    at /home/sinaomrani/mambaforge3/envs/moose/libmesh/include/libmesh/type_vector.h:886
#1  0x00007ffff77f49db in MyNewFluidTempl<false>::initQpStatefulProperties (
    this=0x555556466620)
    at /home/sinaomrani/projects/moose/modules/porous_flow/src/materials/MyNewFluid.C:81
#2  0x00007ffff2a79329 in MaterialBase::initStatefulProperties (
    this=0x555556466620, n_points=4)
    at /home/sinaomrani/projects/moose/framework/src/materials/MaterialBase.C:115
#3  0x00007ffff784b6c0 in PorousFlowMaterial::initStatefulProperties (
    this=0x555556466620, n_points=4)
    at /home/sinaomrani/projects/moose/modules/porous_flow/src/materials/PorousFlowMaterial.C:71
#4  0x00007ffff2a81c3c in MaterialPropertyStorage::initStatefulProps (
    this=0x5555560f2788, material_data=..., mats=..., n_qpoints=4, elem=..., 
    side=0)
    at /home/sinaomrani/projects/moose/framework/src/materials/MaterialPropertyStorage.C:201
#5  0x00007ffff34b5f8a in ComputeMaterialsObjectThread::onElement (
    this=0x7fffffffd1b0, elem=0x5555560d05e0)
--Type <RET> for more, q to quit, c to continue without paging--ret
   /framework/src/loops/ComputeMaterialsObjectThread.C:113
#6  0x00007ffff2f47b59 in ThreadedElementLoopBase<libMesh::StoredRange<libMesh::MeshBase::const_element_iterator, libMesh::Elem const*> >::operator() (this=0x7fffffffd1b0, range=..., 
    bypass_threading=false) at /home/sinaomrani/projects/moose/framework/build/header_symlinks/ThreadedElementLoopBase.h:248
#7  0x00007ffff2f2bfe6 in libMesh::Threads::parallel_reduce<libMesh::StoredRange<libMesh::MeshBase::const_element_iterator, libMesh::Elem const*>, ComputeMaterialsObjectThread> (range=..., body=...)
    at /home/sinaomrani/mambaforge3/envs/moose/libmesh/include/libmesh/threads_pthread.h:380
#8  0x00007ffff2eb0d1b in FEProblemBase::initialSetup (this=0x5555560ee890) at /home/sinaomrani/projects/moose/framework/src/problems/FEProblemBase.C:943
#9  0x00007ffff38cbfcf in Steady::init (this=0x555556147ba0) at /home/sinaomrani/projects/moose/framework/src/executioners/Steady.C:49
#10 0x00007ffff45add56 in MooseApp::executeExecutioner (this=0x5555557b7520) at /home/sinaomrani/projects/moose/framework/src/base/MooseApp.C:1144
#11 0x00007ffff45b24e6 in MooseApp::run (this=0x5555557b7520) at /home/sinaomrani/projects/moose/framework/src/base/MooseApp.C:1445
#12 0x00005555555576a9 in main (argc=3, argv=0x7fffffffdd48) at /home/sinaomrani/projects/moose/modules/porous_flow/src/main.C:33


I would appreciate it if you could point out what is the issue. Thank you for your time.

**Reply by Joseph-0123** [reply]
I guess your MyNewFluid.C has some errors in initQpStatefulProperties settings. It may relate to your old velocity setting.

Please search the related questions about initQpStatefulProperties  https://github.com/idaholab/moose/discussions?discussions_q=is%3Aopen+initQpStatefulProperties

like https://github.com/idaholab/moose/discussions/24291 https://github.com/idaholab/moose/discussions/23263

**Reply by cpgr** [reply]
It might help to paste your code here. At a guess something that you are using in `initQpStatefulProperties` hasn't been sized properly yet.

**Reply by sinaomrani96** [reply]
Hi Joseph,

Thank you for sharing your thoughts. I will look into it in more detail.

Hi Chris,

Sure. Here is the code related to viscosity. It is basically the same as PorousFlowSingleComponentFluid and I only changed the followings:

In MyNewFluid.h file:

`const MaterialProperty<std::vector<RealVectorValue>> & _darcy_velocity;`

and in MyNewFluid.C file:

```
template <bool is_ad>
MyNewFluidTempl<is_ad>::MyNewFluidTempl(
    const InputParameters & parameters)
  : PorousFlowFluidPropertiesBaseTempl<is_ad>(parameters),
    _darcy_velocity(this->template getMaterialPropertyOldByName<std::vector<RealVectorValue>>("PorousFlow_darcy_velocity_qp")),
    _fp(this->template getUserObject<MyFluidProperty>("fp"))
{
}

template <bool is_ad>
void
MyNewFluidTempl<is_ad>::initQpStatefulProperties()
{
  if (_compute_rho_mu)
  {
    (*_density)[_qp] = _fp.rho_from_p_T(_porepressure[_qp][_phase_num] * _pressure_to_Pascals,
                                        _temperature[_qp] + _t_c2k);


    (*_viscosity)[_qp] = _fp.mu_from_p_T(_porepressure[_qp][_phase_num] * _pressure_to_Pascals,
                                         _temperature[_qp] + _t_c2k) /
                         _pressure_to_Pascals / _time_to_seconds;
  }

  if (_compute_internal_energy)
    (*_internal_energy)[_qp] = _fp.e_from_p_T(_porepressure[_qp][_phase_num] * _pressure_to_Pascals,
                                              _temperature[_qp] + _t_c2k);

  if (_compute_enthalpy)
    (*_enthalpy)[_qp] = _fp.h_from_p_T(_porepressure[_qp][_phase_num] * _pressure_to_Pascals,
                                       _temperature[_qp] + _t_c2k);
}


template <bool is_ad>
void
MyNewFluidTempl<is_ad>::computeQpProperties()
{
  Real velocity_abs;
  if (_compute_rho_mu)
  {
    if (is_ad)
    {
      GenericReal<is_ad> rho, mu;
      _fp.rho_mu_from_p_T(_porepressure[_qp][_phase_num] * _pressure_to_Pascals,
                          _temperature[_qp] + _t_c2k,
                          rho,
                          mu);

      (*_density)[_qp] = rho;
      (*_viscosity)[_qp] = mu / _pressure_to_Pascals / _time_to_seconds;
    }
    else
    {
      // Density and viscosity, and derivatives wrt pressure and temperature
      Real rho, drho_dp, drho_dT, mu, dmu_dp, dmu_dT;
      _fp.rho_mu_from_p_T(MetaPhysicL::raw_value(_porepressure[_qp][_phase_num]) *
                              _pressure_to_Pascals,
                          MetaPhysicL::raw_value(_temperature[_qp]) + _t_c2k,
                          rho,
                          drho_dp,
                          drho_dT,
                          mu,
                          dmu_dp,
                          dmu_dT);
      (*_density)[_qp] = rho;
      (*_ddensity_dp)[_qp] = drho_dp * _pressure_to_Pascals;
      (*_ddensity_dT)[_qp] = drho_dT;
      velocity_abs = std::sqrt(_darcy_velocity[_qp][0] * _darcy_velocity[_qp][0]);
      (*_viscosity)[_qp] = velocity_abs;
      (*_dviscosity_dp)[_qp] = 0 / _time_to_seconds;
      (*_dviscosity_dT)[_qp] = 0 / _pressure_to_Pascals / _time_to_seconds;
    }
  }
```

Please don't mind the correlation between viscosity and velocity, I put this only for testing. I appreciate your time. Thanks.

**Reply by cpgr** [reply]
I think that you need to access the velocity vector components with () not [].

In any case you probably could use the vector method to compute this. I think it's just norm() but am not at a computer to check

**Reply by cpgr** [reply]
Perhaps MetaPhysicL::raw_value(_darcy_velocity[_qp]).norm()
